sdk/nodejs/cmd/pulumi-language-nodejs/main.go

```diff
@@ -807,16 +810,76 @@ func (host *nodeLanguageHost) InstallDependencies(
 
 	stdout.Write([]byte("Installing dependencies...\n\n"))
 
-	_, err = npm.Install(ctx, req.Info.ProgramDirectory, false /*production*/, stdout, stderr)
+	workspaceRoot := req.Info.ProgramDirectory
+	newWorkspaceRoot, err := findWorkspaceRoot(req.Info.ProgramDirectory)
 	if err != nil {
-		return fmt.Errorf("npm install failed: %w", err)
+		// If we are not in a npm/yarn workspace, we will keep the current directory as the root.
+		if !errors.Is(err, errNotInWorkspace) {
+			return fmt.Errorf("failure while trying to find workspace root: %w", err)
+		}
+	} else {
+		stdout.Write([]byte(fmt.Sprintf("Detected workspace root at %s\n", newWorkspaceRoot)))
+		workspaceRoot = newWorkspaceRoot
+	}
+
+	_, err = npm.Install(ctx, workspaceRoot, false /*production*/, stdout, stderr)
+	if err != nil {
+		return fmt.Errorf("dependency installation failed: %w", err)
 	}
 
 	stdout.Write([]byte("Finished installing dependencies\n\n"))
 
 	return closer.Close()
 }
 
+func findWorkspaceRoot(programDirectory string) (string, error) {
+	currentDir := filepath.Dir(programDirectory)
+	nextDir := filepath.Dir(currentDir)
+	for currentDir != nextDir { // We're at the root when the nextDir is the same as the currentDir.
+		p := filepath.Join(currentDir, "package.json")
+		_, err := os.Stat(p)
+		if err != nil {
+			if os.IsNotExist(err) {
+				// No package.json in this directory, continue the search in the next directory up.
+				currentDir = nextDir
+				nextDir = filepath.Dir(currentDir)
+				continue
+			}
+			return "", err
+		}
+
+		// Get the workspaces field from the package.json
+		pkgContents, err := os.ReadFile(p)
+		if err != nil {
+			return "", err
+		}
+		pkg := struct {
+			Workspaces []string `json:"workspaces"`
+		}{}
+		err = json.Unmarshal(pkgContents, &pkg)
+		if err != nil {
```

Comment:
Should this be a hard error? If someone has a "package.json" file with malformed JSON or a "workspaces" field set to something else, should we just assume this isn't a real package.json and keep searching up?

Reply:
Good question. I  checked the behaviour of yarn and npm just now:

For npm it fails immediately with:
```
npm ERR! code EWORKSPACESCONFIG
npm ERR! workspaces config expects an Array
```

For yarn it starts trying to install, but then fails because it can't find the other workspace packages.

However while looking at this I saw that yarn also supports an alternative config here:

```
 "workspaces": {
    "packages": ["packages/*"],
    "nohoist": ["**/react-native", "**/react-native/**"]
  }
```

So I will need to update the code to support his.

I think we can hard error if we can't parse either format though.